server/schedule/checker/rule_checker.go

```diff
@@ -266,6 +279,24 @@ func (c *RuleChecker) fixLooseMatchPeer(region *core.RegionInfo, fit *placement.
 		checkerCounter.WithLabelValues("rule_checker", "demote-voter-role").Inc()
 		return operator.CreateDemoteVoterOperator("fix-demote-voter", c.cluster, region, peer)
 	}
+	if region.GetLeader().GetId() == peer.GetId() && rf.Rule.IsWitness {
```

Comment:
I think we need to make sure a learner with a witness cannot be promoted.

Reply:
I think
- {A(voter), B(voter), C(learner, witness)} + AddNode(C(witness)) = {A(voter), B(voter), C(voter, witness)}

is legal.


Are you referring to this conversion path?

Reply:
The current implementation guarantees this, first scheduling the conversion between the learner and the follower, and then doing the witness conversion scheduling